File: src/04_quicksort/python/05_quicksort.py
def quicksort(arr: list[int]) -> int:
    if len(arr) < 2:
        # base case, arrays with 0 or 1 element are already "sorted"
        return arr
    else:
        # recursive case
        pivot = arr[0]
        # sub-array of all the elements less than the pivot
        less = [i for i in arr[1:] if i <= pivot]
        # sub-array of all the elements greater than the pivot
        greater = [i for i in arr[1:] if i > pivot]
        # less subarray + pivot + greater subarray
        return quicksort(less) + [pivot] + quicksort(greater)


print(quicksort([10, 5, 2, 3, 5, 6, 9, 8, 7, 5, 2, 5, 8]))

File: src/04_quicksort/ts/05_quicksort.ts
function partiton(arr: number[], pivot: number, subarr: 'right' | 'left') {
  if (subarr === 'right') {
    const subArray = arr.slice(1).filter((item) => item >= pivot);
    return subArray;
  } else {
    const subArray: number[] = [];
    for (let i = 0; i < arr.length; i++) {
      if (arr[i] <= pivot) {
        subArray.push(arr[i]);
      }
      console.log(arr[i]);
    }
    return subArray;
  }
}

function quicksort(arr: number[]): number[] {
  // base case, arrays with 0 or 1 element are already "sorted"
  if (arr.length < 2) {
    return arr;
  } else {
    // recursive case
    const pivot = arr[0];
    // sub-array of all the elements less than the pivot
    const less: number[] = arr.slice(1).filter((item) => item <= pivot);
    // sub-array of all the elements greater than the pivot
    const greater: number[] = arr.slice(1).filter((item) => item >= pivot);
    // less subarray + pivot + greater subarray
    return quicksort(less).concat([pivot], quicksort(greater));
  }
}

console.log(quicksort([5, 5, 3, 6, 2]));
